substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -1704,6 +1815,381 @@ impl<T: Config> Pallet<T> {
+            matches!(
+                service_contract.state,
+                types::ServiceContractState::AgreementReady
+            ),
+            Error::<T>::ServiceContractRejectionNotAllowed,
+        );
+
+        // Only service or consumer can reject agreement
+        ensure!(
+            twin_id == service_contract.service_twin_id
+                || twin_id == service_contract.consumer_twin_id,
+            Error::<T>::TwinNotAuthorized,
+        );
+
+        // If one party (service or consumer) rejects agreement
+        // then contract is canceled and removed from service contract map
+        Self::_service_contract_cancel(
+            account_id,
+            service_contract_id,
+            types::Cause::CanceledByUser,
+        )?;
+
+        Ok(().into())
+    }
+
+    pub fn _service_contract_cancel(
+        account_id: T::AccountId,
+        service_contract_id: u64,
+        cause: types::Cause,
+    ) -> DispatchResultWithPostInfo {
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        let service_contract = ServiceContracts::<T>::get(service_contract_id)
+            .ok_or(Error::<T>::ServiceContractNotExists)?;
+
+        // Only service or consumer can cancel contract
+        ensure!(
+            twin_id == service_contract.service_twin_id
+                || twin_id == service_contract.consumer_twin_id,
+            Error::<T>::TwinNotAuthorized,
+        );
+
+        // Remove contract from service contract map
+        // Can be done at any state of contract
+        // so no need to handle state validation
+        ServiceContracts::<T>::remove(service_contract_id);
+        Self::deposit_event(Event::ServiceContractCanceled {
+            service_contract_id,
+            cause,
+        });
+
+        log::info!(
+            "successfully removed service contract with id {:?}",
+            service_contract_id,
+        );
+
+        Ok(().into())
+    }
+
+    #[transactional]
+    pub fn _service_contract_bill(
+        account_id: T::AccountId,
+        service_contract_id: u64,
+        variable_amount: u64,
+        metadata: Vec<u8>,
+    ) -> DispatchResultWithPostInfo {
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        let mut service_contract = ServiceContracts::<T>::get(service_contract_id)
+            .ok_or(Error::<T>::ServiceContractNotExists)?;
+
+        // Only service can bill consumer for service contract
+        ensure!(
+            twin_id == service_contract.service_twin_id,
+            Error::<T>::TwinNotAuthorized,
+        );
+
+        // Allow to bill contract only if approved by both
+        ensure!(
+            matches!(
+                service_contract.state,
+                types::ServiceContractState::ApprovedByBoth
+            ),
+            Error::<T>::ServiceContractBillingNotAllowed,
```

Comment:
This error doesn't really tell me what went wrong.... Maybe add another error saying that both should approve?